### DISCUSSION TITLE
How to gracefully close a connection

### DISCUSSION THREAD

**Discussion by chris13524** [tokio-rs/axum#2115] [Q&A] [answered]
How can I close a connection after a certain timeout?

My issue is that I have a lot of browser keep-alive/H2 connections that are using up a lot of memory. I know this because disabling [HTTP/1 keep-alive](https://docs.rs/hyper/0.14.27/hyper/server/struct.Builder.html#method.http1_keepalive) greatly reduces the memory utilization.

Currently I am doing this using a custom executor with [`hyper::Server::executor`](https://docs.rs/hyper/0.14.27/hyper/server/struct.Builder.html#method.executor) and canceling the future with `Future::with_timeout` but I'm concerned this isn't the right approach as I don't think it utilizes [`Connection::graceful_shutdown`](https://docs.rs/hyper/0.14.27/hyper/server/conn/struct.Connection.html#method.graceful_shutdown)

**Comment by davidpdrsn** [comment, marked as answer]
The easiest way is probably using hyper low level APIs. That gives good control over stuff like this.

For more details I recommend you ask in the hyper repo or discord channel. axum itself isn’t really concerned with the connection

**Reply by chris13524** [reply]
I think you're totally right. I need to use the [`conn`](https://docs.rs/hyper/latest/hyper/server/conn/index.html) module.
